client_test.go

```diff
@@ -199,6 +199,100 @@ func TestClientMetadata(t *testing.T) {
 	safeClose(t, client)
 }
 
+func TestClientReceivingUnknownTopic(t *testing.T) {
+	seedBroker := newMockBroker(t, 1)
+
+	metadataResponse1 := new(MetadataResponse)
+	metadataResponse1.AddBroker(seedBroker.Addr(), seedBroker.BrokerID())
+	seedBroker.Returns(metadataResponse1)
+
+	config := NewConfig()
+	config.Metadata.Retry.Max = 0
+	client, err := NewClient([]string{seedBroker.Addr()}, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	metadataResponse2 := new(MetadataResponse)
+	metadataResponse2.AddTopic("new_topic", ErrUnknownTopicOrPartition)
+	seedBroker.Returns(metadataResponse2)
+
+	if err := client.RefreshMetadata("new_topic"); err != ErrUnknownTopicOrPartition {
+		t.Error("ErrUnknownTopicOrPartition expected, got", err)
+	}
+
+	// If we are asking for the leader of a partition of the non-existing topic.
+	// we will request metadata again.
+
+	metadataResponse3 := new(MetadataResponse)
+	metadataResponse3.AddTopic("new_topic", ErrUnknownTopicOrPartition)
+	seedBroker.Returns(metadataResponse3)
```

Comment:
you can just re-use `metadataResponse2` here, there's no need to construct a new one that's exactly the same